isabelle/Core/SemanticsTypes.thy

```diff
@@ -367,29 +258,79 @@ and Contract = Close
+ we follow the continuation associated to the first \<^term>\<open>Action\<close> that matches the \<^term>\<open>Input\<close>. If no
+ action is matched it returns a \<^term>\<open>ApplyAllNoMatchError\<close>. If a valid \<^term>\<open>Transaction\<close> is computed
+ with a \<^term>\<open>TimeInterval\<close> with a start time bigger than the \<^term>\<open>Timeout t\<close>, the contingency
+ continuation \<^term>\<open>c\<close> is evaluated. The explicit timeout mechanism is what allow Marlowe to avoid
+waiting forever for external inputs.\<close>
+
+text \<open>A @{term Let} contract @{term "Let i v c"} allows a contract to name a
+value using an identifier @{term i}. In this case, the expression @{term v} is
+evaluated, and stored with the name @{term i}. The contract then continues as
+@{term c}. As well as allowing us to use abbreviations, this mechanism also means that we
+can capture and save volatile values that might be changing with time, e.g. the current price of
+oil, or the current time, at a particular point in the execution of the contract, to be used later
+on in contract execution.
+\<close>
+
+text \<open>An assertion contract @{term "Assert b c"} does not have any effect on
+the state of the contract, it immediately continues as @{term c}, but it issues a warning
+when the observation @{term b} is false. It can be used to ensure that a property holds
+in any given point of the contract, since static analysis will fail if any execution causes a
+warning. the @{term Assert} term might be removed from future on-chain versions of Marlowe.\<close>
+
+
+subsection \<open>State and Environment \label{sec:state-and-env}\<close>
+
+text \<open>The internal state of a Marlowe contract consists of the current balances in each party's
+account, a record of the most recent value of each type of choice, a record of the most recent
+value of each variable, and the minimum time for which input can be applied to the contract. The
+data for accounts, choices, and bound values are stored as association lists.\<close>
 
 record State = accounts :: Accounts
                choices :: "(ChoiceId \<times> ChosenNum) list"
                boundValues :: "(ValueId \<times> int) list"
                minTime :: POSIXTime
 
-fun valid_state :: "State \<Rightarrow> bool" where
-"valid_state state = (valid_map (accounts state)
-                     \<and> valid_map (choices state)
-                     \<and> valid_map (boundValues state))"
+text \<open>The execution environment of a Marlowe contract simply consists of the (inclusive) time
+interval within which the transaction is occurring.\<close>
 
 record Environment = timeInterval :: TimeInterval
 
-datatype Input = IDeposit AccountId Party Token Money
-               | IChoice ChoiceId ChosenNum
-               | INotify
 
+
+
+
+\<comment> \<open>TODO: see if we want to add data types of Semantic here (Transaction, etc) or if we want to
```

Comment:
Perhaps we could collect all of the error and warning types in a new section here?